virtcontainers/pkg/types/pcipath.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2020 David Gibson, Red Hat Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package types
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+const maxPciSlot = 31
+
+// A PciPath describes where a PCI sits in a PCI hierarchy.
+//
+// Consists of a list of PCI slot numbers, giving the slot of each
+// bridge that must be traversed from the PCI root to reach the
+// device, followed by the slot of the device itself
+//
+// When formatted into a string is written as "xx/.../yy/zz" Here, zz
+// is the slot of the device on its PCI bridge, yy is the slot of the
+// bridge on its parent bridge and so forth until xx is the slot of
+// the "most upstream" bridge on the root bus.  If a device is
+// connected directly to the root bus, its PciPath is just "zz"
+
+type PciPath struct {
+	path []uint8
+}
+
+func (p PciPath) Format() string {
+	tokens := make([]string, len(p.path))
+	for i, slot := range p.path {
+		tokens[i] = fmt.Sprintf("%02x", slot)
+	}
+	return strings.Join(tokens, "/")
+}
+
+func (p PciPath) IsNil() bool {
+	return p.path == nil
+}
+
+func PciPathFromString(s string) (PciPath, error) {
+	if s == "" {
+		return PciPath{}, nil
+	}
+
+	tokens := strings.Split(s, "/")
+	slots := make([]uint8, len(tokens))
+	for i, t := range tokens {
+		slot, err := strconv.ParseUint(t, 16, 5)
+		if err != nil {
+			return PciPath{}, err
```

Comment:
I'd be tempted to mark this as a bug since if the PCI path was already validated in `MakePciPath()`, this should be "impossible" to fail:

```go
return PciPath{}, fmt.Errorf("BUG: %q resulted in invalid PCI slot %v", s, err)
```

Reply:
The input to this is a plain string though, it need not have come from MakePciPath().  So it's not necessarily a bug... or at least, not necessarily a bug in this component.